File: 프로그래머스/월간 코드 챌린지 시즌2/110 옮기기/110 옮기기.cpp
#include <string>
#include <vector>
using namespace std;

vector<string> solution(vector<string> s) {
    vector<string> ans;
    for (string si : s) {
        vector<char> st;
        int cnt110 = 0;
        for (int i = 0; i < si.length(); i++) {
            // st 크기가 2 미만이면 push
            if (st.size() < 2) st.push_back(si[i]);
            // 1은 push
            else if (si[i] == '1') st.push_back('1');
            else {
                // st에 연달아 11이 있으면 빼고 카운팅
                if (st[st.size() - 1] == '1' && st[st.size() - 2] == '1') {
                    st.pop_back(); st.pop_back();
                    cnt110++;
                }
                else st.push_back('0');
            }
        }

        // 1이 두 번 연달아 나타나는 곳 앞에 110 붙이기
        string answer = "";
        int idx = 0;
        for (; idx < st.size(); idx++) {
            if (st[idx] == '0') answer += "0";
            else {
                if (idx == st.size() - 1 || st[idx + 1] == '1') {
                    for (int j = 0; j < cnt110; j++)
                        answer += "110";
                    for (int j = idx; j < st.size(); j++)
                        answer += st[j];

                    cnt110 = 0;
                    break;
                }
                else answer += "1";
            }
        }

        if (cnt110)
            for (int i = 0; i < cnt110; i++)
                answer += "110";

        ans.push_back(answer);
    }

    return ans;
}

File: 프로그래머스/월간 코드 챌린지 시즌2/110 옮기기/110 옮기기.py
def solution(s):
    ans = []
    for si in s:
        st = []
        cnt110 = 0
        for i in range(len(si)):
            if len(st) < 2: st.append(si[i])
            elif si[i] == '1': st.append('1')
            else:
                if st[-1]=='1' and st[-2]=='1':
                    st.pop(-1); st.pop(-1)
                    cnt110 += 1
                else: st.append('0')

        answer = ""
        for i in range(len(st)):
            if st[i] == '0': answer += '0'
            else:
                if i==len(st)-1 or st[i+1]=='1':
                    for j in range(cnt110):
                        answer += "110"
                    for j in range(i, len(st)):
                        answer += st[j]

                    cnt110 = 0
                    break
                else: answer += '1'

        if cnt110 > 0:
            for i in range(cnt110):
                answer += "110"

        ans.append(answer)

    return ans
